kv/txn_coord_sender.go

```diff
@@ -802,3 +783,61 @@ func (tc *TxnCoordSender) heartbeat(id string) {
 		}
 	}
 }
+
+func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.Context) bool {
+	tc.Lock()
+	proceed := true
+	txnMeta := tc.txns[id]
+	// Before we send a heartbeat, determine whether this transaction
+	// should be considered abandoned. If so, exit heartbeat.
+	if txnMeta.hasClientAbandonedCoord(tc.clock.PhysicalNow()) {
+		// TODO(tschottdorf): should we be more proactive here?
+		// The client might be continuing the transaction
+		// through another coordinator, but in the most likely
+		// case it's just gone and the open transaction record
+		// could block concurrent operations.
+		if log.V(1) {
+			log.Infof("transaction %s abandoned; stopping heartbeat",
+				txnMeta.txn)
+		}
+		proceed = false
+	}
+	// txnMeta.txn is possibly replaced concurrently,
+	// so grab a copy before unlocking.
+	txn := txnMeta.txn
+	tc.Unlock()
+	if !proceed {
+		tc.unregisterTxn(id)
```

Comment:
This method is mostly just code movement, but there are two changes. This line was added...